meson: the curses flavour is chosen in three places that can disagree

> Drafted by Claude (Opus 5), an LLM made by Anthropic, and posted at the
> direction of andy5995.

## The point

The meson build picks a curses flavour — wide or narrow — in three places that
can disagree: the curses dependency, the menu dependency, and the header probe.
Nothing ties them together, so a build can pair the wide menu library with narrow
curses, or compile against the wide header while linking the narrow library.
Either combination links cleanly.

## Where they diverge

**The library pair.** `dep_curses` is resolved generically, while the menu lookup
tries `menuw` first and falls back to `menu`. Those searches are independent, so
`menuw` can be selected alongside narrow `ncurses`, or `menu` alongside
`ncursesw`. The menu library that ships with ncurses is built against one flavour
and shares its `WINDOW`/`chtype` expectations, so a crossed pair is a silent
mismatch rather than a link error.

**The header.** The `cc.has_header` loop sets `HAVE_NCURSESW_CURSES_H` purely
because `ncursesw/curses.h` exists on the build machine, and the `#if`-cascade in
the sources then prefers it. That choice never consults which dependency was
resolved. `ncursesw/curses.h` defines `NCURSES_WIDECHAR` as 1, so it declares
`cchar_t` and the `*_wch` family — potentially against a library that does not
provide them.

The general rule behind both: `cc.has_header` answers "does this file exist", not
"is this the header for the library I am linking".

## What should not change

The layered menu lookup is there for a reason, and any fix needs to keep it. The
ChangeLog entry for 2022-01-25 (PR #334) records it: "the curses menu library is
now much more likely to be found on \*BSD systems". The `find_library` fallbacks
exist because pkg-config cannot be relied on to find the menu library there — on
this repo's own BSD CI, neither leg installs ncurses, and the OpenBSD leg does not
install pkgconf either. On Arch, by contrast, `ncurses.pc`, `ncursesw.pc`,
`menu.pc` and `menuw.pc` are all present, so pkg-config alone is sufficient and
the fallbacks never run.

## Suggested direction

Decide the flavour once, then look for that flavour's menu library through the
same fallback chain, instead of searching wide-then-narrow independently:

1. try `dependency('ncursesw', required: false)`; if found, the flavour is wide,
   otherwise fall back to `ncurses`;
2. for the chosen flavour, look for its menu library by
   `dependency()` → `cc.find_library()`, keeping the existing BSD-friendly
   fallbacks but only for that one name;
3. set the header define from the resolved dependency rather than from bare
   existence, e.g. by passing it to the probe:
   `cc.has_header('ncursesw/curses.h', dependencies: dep_curses)`.

`menucd` already uses the first shape for its own `ncursesw` and `menuw` lookups.

## Not verified

I found this by reading the build files and have not reproduced either failure;
no system is confirmed to hit it. On Arch it cannot happen, because there is no
`/usr/include/ncursesw/` directory at all, so the cascade falls through to
`curses.h` — which is itself the wide header there — and all three choices agree
by coincidence. Whether any \*BSD or older Debian layout produces a genuine
mismatch is exactly the part that would need checking on those systems.

I would treat this as latent rather than urgent.

> Drafted by Claude (Opus 5), an LLM made by Anthropic, and posted at the
> direction of andy5995.

Correcting a claim in the description, and adding what the sources actually say.

**The menu lookup is load-bearing on every platform, not a \*BSD workaround.** I
tested by removing it from a throwaway copy of the build on Arch: the link then
fails with `undefined reference to item_name`, `new_item` and friends. The
pkg-config metadata shows why — the dependency runs from menu towards curses, so
finding curses can never bring menu with it:

```
ncursesw.pc:  Libs: -lncursesw
menuw.pc:     Libs: -lmenuw    Requires.private: ncursesw
```

**So the reason for the `find_library` fallbacks is probably not missing `.pc`
files.** I guessed that in the description; the OpenBSD sources say otherwise.
In `lib/libmenu/Makefile`, OpenBSD base installs **both** `menu.pc` and
`menuw.pc`, and `lib/libform` and `lib/libpanel` do the same for their pairs. A
likelier explanation is that pkg-config itself may be absent: this repo's OpenBSD
CI leg installs `gettext-tools git glib2 meson ninja` and no pkgconf, in which
case `dependency()` cannot work at all while `cc.find_library()` still can. That
part I have not confirmed on a running \*BSD.

**Also relevant to the suggested fix:** OpenBSD's base curses does have the wide
API, so preferring `ncursesw` is safe there. `lib/libcurses/Makefile` builds with
`-D_XOPEN_SOURCE_EXTENDED`, includes a `widechar` source directory, and installs
both `ncurses.pc` and `ncursesw.pc` for the one `libcurses`. That is worth knowing
because `_XOPEN_SOURCE_EXTENDED` is precisely what gates `cchar_t` and the `*_wch`
family in `curses.h`; a platform that does not define it compiles the narrow
fallback even with a wide-capable library installed.

None of this changes the two problems in the description — the flavour is still
chosen in three places that can disagree. It does mean any fix has to keep looking
for the menu library explicitly, and that deriving it from the chosen flavour is
viable on OpenBSD as well.

--- a/ChangeLog
+++ b/ChangeLog
@@ -1,5 +1,8 @@
 
 rmw (in-progress):
+  * build: pick the curses and menu libraries so they cannot disagree, and
+    probe the curses headers against the library that was found rather than
+    against whatever exists on the system (#579)
   * verbose: the startup path/diagnostic lines (home_dir, config_file,
     discovered topdir trashes, ficlone support, etc.) now require -vv; plain
     -v keeps only the per-file action messages

--- a/meson.build
+++ b/meson.build
@@ -44,23 +44,6 @@ conf.set_quoted(
   'https://github.com/theimpossibleastronaut/rmw/issues',
 )
 
-check_headers = [
-  ['ncursesw/menu.h', 'HAVE_NCURSESW_MENU_H'],
-  ['ncurses/menu.h', 'HAVE_NCURSES_MENU_H'],
-  ['menu.h', 'HAVE_MENU_H'],
-  ['ncursesw/curses.h', 'HAVE_NCURSESW_CURSES_H'],
-  ['ncursesw.h', 'HAVE_NCURSESW_H'],
-  ['ncurses/curses.h', 'HAVE_NCURSES_CURSES_H'],
-  ['ncurses.h', 'HAVE_NCURSES_H'],
-  ['curses.h', 'HAVE_CURSES_H'],
-]
-
-foreach h : check_headers
-  if cc.has_header(h[0])
-    conf.set(h[1], 1)
-  endif
-endforeach
-
 localedir = join_paths(get_option('prefix'), get_option('localedir'))
 subdir('po')
 

--- a/src/meson.build
+++ b/src/meson.build
@@ -3,23 +3,63 @@ disable_curses = ''
 dep_menu = []
 dep_curses = []
 if not get_option('without-curses')
+  # Meson's curses dependency has a SYSTEM method of its own, so this resolves
+  # even where no pkg-config files exist for it.
   dep_curses = dependency('curses')
+
+  # The menu library ships with ncurses but is separate to the linker, and meson
+  # has no menu dependency, so it has to be looked for by hand. The order is
+  # deliberate:
+  #
+  #   menuw via pkg-config    matches the wide curses meson prefers, and is the
+  #                           only route that finds a keg-only Homebrew ncurses
+  #   the plain library names  need no pkg-config data at all; OpenBSD shipped
+  #                           none for these before 2025, and its single libmenu
+  #                           is already wide
+  #   dependency('menu') last  menu.pc names the narrow library in
+  #                           Requires.private, so it can pull a second curses
+  #                           into the link (mesonbuild/meson#8867)
   dep_menu = dependency('menuw', required: false)
   if not dep_menu.found()
+    dep_menu = cc.find_library('menuw', required: false)
+  endif
+  if not dep_menu.found()
+    dep_menu = cc.find_library('menu', required: false)
+  endif
+  if not dep_menu.found()
     dep_menu = dependency('menu', required: false)
-    if not dep_menu.found()
-      dep_menu = cc.find_library('menuw')
-      if not dep_menu.found()
-        dep_menu = cc.find_library('menu')
-        if not dep_menu.found()
-          conf.set('DISABLE_CURSES', 1)
-        endif
-      endif
-    endif
+  endif
+  if not dep_menu.found()
+    conf.set('DISABLE_CURSES', 1)
   endif
 else
   conf.set('DISABLE_CURSES', 1)
 endif
+
+# Probe each header against the curses that was actually found, rather than just
+# for the file being present. The HAVE_* names below are borrowed from
+# AX_WITH_CURSES, an autoconf-archive m4 macro that this project does not use and
+# does not ship -- only its define names were taken:
+#   https://gitweb.git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_with_curses.m4
+# There each one is set by a link test against the library the macro selected, so
+# it means "this header works with our curses". Passing dep_curses keeps that
+# meaning here. It is empty when curses is disabled, so that case is unchanged.
+check_headers = [
+  ['ncursesw/menu.h', 'HAVE_NCURSESW_MENU_H'],
+  ['ncurses/menu.h', 'HAVE_NCURSES_MENU_H'],
+  ['menu.h', 'HAVE_MENU_H'],
+  ['ncursesw/curses.h', 'HAVE_NCURSESW_CURSES_H'],
+  ['ncursesw.h', 'HAVE_NCURSESW_H'],
+  ['ncurses/curses.h', 'HAVE_NCURSES_CURSES_H'],
+  ['ncurses.h', 'HAVE_NCURSES_H'],
+  ['curses.h', 'HAVE_CURSES_H'],
+]
+
+foreach h : check_headers
+  if cc.has_header(h[0], dependencies: dep_curses)
+    conf.set(h[1], 1)
+  endif
+endforeach
 
 src = [
   'globals.c',

